es-core/src/math/Matrix4x4.h

```diff
@@ -0,0 +1,160 @@
+#define _MATRIX4X4_H_
+
+#include <assert.h>
+#include <math/Vector4.h>
+#include <math/Vector3.h>
+
+template<typename S>
+class Matrix4x4
+{
+public:
+
+	typedef Matrix4x4<S> M;
+
+	Matrix4x4() { }
+	Matrix4x4(const Vector4<S> x, const Vector4<S> y, const Vector4<S> z, const Vector4<S> w) : mX(x), mY(y), mZ(z), mW(w) { };
+
+	const bool operator==(const M& other) const { return ((mX == other.x()) && (mY == other.y()) && (mZ == other.z()) && (mW == other.w())); }
+	const bool operator!=(const M& other) const { return ((mX != other.x()) || (mY != other.y()) || (mZ != other.z()) || (mW != other.w())); }
+
+	const M operator+(const M& other) const { return { mX + other.x(), mY + other.y(), mZ + other.z(), mW + other.w() }; }
+	const M operator-(const M& other) const { return { mX - other.x(), mY - other.y(), mZ - other.z(), mW - other.w() }; }
+
+	const M operator*(const M& other) const
+	{
+		return
+		{
+			{ other.x() * mX.x() + other.y() * mX.y() + other.z() * mX.z() + other.w() * mX.w() },
+			{ other.x() * mY.x() + other.y() * mY.y() + other.z() * mY.z() + other.w() * mY.w() },
+			{ other.x() * mZ.x() + other.y() * mZ.y() + other.z() * mZ.z() + other.w() * mZ.w() },
+			{ other.x() * mW.x() + other.y() * mW.y() + other.z() * mW.z() + other.w() * mW.w() },
+		};
+	}
+
+	const Vector3<S> operator*(const Vector3<S>& other) const
+	{
+		return
+		{
+			other.x() * mX.x() + other.y() * mY.x() + other.z() * mZ.x() + mW.x(),
+			other.x() * mX.y() + other.y() * mY.y() + other.z() * mZ.y() + mW.y(),
+			other.x() * mX.z() + other.y() * mY.z() + other.z() * mZ.z() + mW.z(),
+		};
+	}
+
+	M& operator+=(const M& other) { *this = *this + other; return *this; }
+	M& operator-=(const M& other) { *this = *this - other; return *this; }
+	M& operator*=(const M& other) { *this = *this * other; return *this; }
+
+	S& operator[](const size_t index) { assert(index < Size() && "index out of range"); return mM[index]; }
+	const S& operator[](const size_t index) const { assert(index < Size() && "index out of range"); return mM[index]; }
+
+	CommaInitializer<M, S> M::operator<<(const S& s) { return CommaInitializer<M, S>(*this, s); }
+
+	Vector4<S>& x() { return mX; }
+	Vector4<S>& y() { return mY; }
+	Vector4<S>& z() { return mZ; }
+	Vector4<S>& w() { return mW; }
+	const Vector4<S>& x() const { return mX; }
+	const Vector4<S>& y() const { return mY; }
+	const Vector4<S>& z() const { return mZ; }
+	const Vector4<S>& w() const { return mW; }
+
+	S* data() { return (S*)this; }
+	const S* data() const { return (S*)this; }
+
+	Vector3<S>& translation() { return mW.v3(); }
+	const Vector3<S>& translation() const { return mW.v3(); }
+
+	M& scale(const Vector3<S>& scale)
+	{
+		mX.v3() *= scale;
+		mY.v3() *= scale;
+		mZ.v3() *= scale;
+		mW.v3() *= scale;
+
+		return *this;
+	}
+
+	M& rotate(const S angle, const Vector3<S>& axis)
+	{
+		const S pi = (S)3.1415926535897932384626433832795028841971693993751058209749445923;
+		const S	a  = -angle * (pi / 180);
```

Comment:
I believe Eigen's rotate method expected angles to already be in radians and as such that is what we are already passing in.

Reply:
Should Pi be moved to a static class member considering that it will never change?